src/onnx/parse_if.cpp

```diff
@@ -65,13 +89,102 @@ struct parse_if : op_parser<parse_if>
 
         auto then_out_shapes = then_mdl->get_output_shapes();
         auto else_out_shapes = else_mdl->get_output_shapes();
-        if(not std::equal(then_out_shapes.begin(),
-                          then_out_shapes.end(),
-                          else_out_shapes.begin(),
-                          else_out_shapes.end()))
-        {
+
+        auto throw_shapes = [&]() {
             MIGRAPHX_THROW("PARSE_IF: " + info.name +
-                           " then and else sub_grahps must have same output shapes!");
+                           " then and else sub_graphs must have compatible shapes ");
+        };
+
+        if(then_out_shapes.size() != else_out_shapes.size())
+        {
+            throw_shapes();
+        }
+
+        // Add checks for each output shape
+        for(int i = 0; i < then_out_shapes.size(); i++)
+        {
+            const auto& then_out_shape = then_out_shapes.at(i);
+            const auto& else_out_shape = else_out_shapes.at(i);
+
+            // Must have the same type for both if/else blocks by onnx spec
+            if(then_out_shape.type() != else_out_shape.type())
+            {
+                MIGRAPHX_THROW("PARSE_IF: " + info.name +
+                               " then and else sub_grahps must have same output type! " +
+                               then_out_shape.type_string() + " vs " +
+                               else_out_shape.type_string());
+            }
+
+            if(not then_out_shape.dynamic() and not else_out_shape.dynamic())
```

Comment:
It would be better to do:

```cpp
if(then_out_shape.dynamic() or else_out_shape.dynamic())
    continue;
```

And avoid extra nesting.

Reply:
Got it. Done.